### ISSUE TITLE
KinematicBody2D is jittering when it's position is set relative to a scrolling camera.

### ISSUE THREAD

**Issue by JosiahJohnson** [godotengine/godot#49718]
[Jitter Test.zip](https://github.com/godotengine/godot/files/6680991/Jitter.Test.zip)
### Godot version

3.3.1

### System information

Win 10, GLES3, i5 4570, RX 580

### Issue description

I have a vertical autoscrolling stage where I'm moving the camera position with a tween. I'm also setting a kinematicbody2D's position relative to the moving camera in _physics_process. This causes the kinematicbody2D to jitter vertically, when it should appear still.  Sync To Physics is also enabled on the kinematicbody2D.

This used to work correctly, but the issue started when I updated to version 3.3.1 from 3.2.3. After testing various versions, the issue starts on 3.2.4 rc4. Possibly related to the 2D snap feature that was scrapped, mentioned in the [RC4 notes](https://godotengine.org/article/release-candidate-godot-3-2-4-rc-4).

Additional info can be found [here](https://www.reddit.com/r/godot/comments/o2qw26/raft_is_now_jittery_after_updating_to_godot_331/).

### Steps to reproduce

Camera code:
`$ScrollTween.interpolate_property(camera, "position", camera.position, stopPosition, scroll_time, Tween.TRANS_LINEAR, Tween.EASE_IN_OUT)`
`$ScrollTween.start()`

kinematicbody2D code:
`position = Vector2(position.x + direction, camera.position.y + float_y)`

**Comment by Calinou** [comment]
Can you test 3.3.2?

**Comment by JosiahJohnson** [comment]
Just tested on 3.3.2, it still jitters.

**Comment by lawnjelly** [comment]
The snapping was reverted in #46657, so that should be the same as in previous version - I'll double check though that there are no differences, it's a while since I looked at it.

What did change is a bug was fixed which was responsible for a frame delay in the 2d camera : #46697. I did suggest making a compatibility mode with the previous version but reviewers preferred not to have this option. If there is something in your code that relied on the previous buggy behaviour, particularly in regard to order of operations, then you could see differences.

Also any changes to physics could also have affected this.

**Comment by JosiahJohnson** [comment]
I'll see if I can throw a test project together.

**Comment by JosiahJohnson** [comment]
I made a test project in 3.3.2 and it scrolls perfectly! I'll have to do some digging to figure out the issue with my game. It might just be a setting a missed.

**Comment by golddotasksquestions** [comment]
What I do in these situations is to duplicate everything (the complete project folder) and import is as a separate project, then delete everything bit by bit that is not essential to replicate the issue.

This takes a bit of time as you have to test to see if the issue still exists, but imho it's still the fastest way if asking the community and recreating from scratch failed.

**Comment by JosiahJohnson** [comment]
Sounds good, I'll have to give that a try.

I stumbled upon an even bigger issue while testing it on my laptop though. My game is completely glitchy on 3.3.2 on my laptop, while it works fine on my desktop. Maybe windows 7 support wasn't thoroughly tested on the new release? It works fine with 3.2.3 on my laptop. I can submit a second issue for that though.

**Comment by Zireael07** [comment]
Windows version shouldn't matter - more likely it's your laptop GPU causing the glitches - and yes, separate issue for those.

**Comment by JosiahJohnson** [comment]
I take that back, I was able to reproduce the jitter issue on my test project (it's attached at the top). I forgot I had **Sync To Physics** enabled on the kinematic body. Once I checked that, it will jitter.

It only occurs on my desktop though (specs listed at the top). Funny enough, it runs fine on my old laptop that has the other issues with my game on the newest version.

**Comment by lawnjelly** [comment]
I think there's several problems going on here:

1) If you put your physics tick rate down to 10tps you will see the effect exaggerated.
2) You are updating your raft on a physics tick, and the camera on a frame process, this might be leading to issues between the two.
3) If you turn off `sync to physics` and change the raft update to `_process` the jitter goes away. I'm not exactly sure what sync to physics does, I'll try and find out.

**Comment by JosiahJohnson** [comment]
Thanks for the info, I didn't know about testing at different tick rates. That's the setting under Physics > Common > Physics Fps? I had sync to physics on all my moving platforms so the player won't slide off, but I don't think it matters in this case since I'm setting the position? My other platforms use tweens for movement, so I'll give that a try.

**Comment by Calinou** [comment]
> That's the setting under Physics > Common > Physics Fps?

Yes :slightly_smiling_face:

**Comment by lawnjelly** [comment]
BTW if you are finding this kind of jitter (player against camera, objects against background) difficult, it is because it *is* much more difficult than it appears, especially when using a physics engine.
I wrote this recently which may give you some sense of the problems involved:
https://github.com/lawnjelly/godot-snapping-demo

**Comment by JosiahJohnson** [comment]
Thanks, interesting read. I had assumed old games worked that way. I’m actually not too picky on perfect scrolling. Funny enough, I find a little jank can give some character. Kinda like warped textures on ps1 games. The raft was a little too much character though 😄

So are there plans to fix issues like this, or is it not possible? It seems like a common scenario others might try, at least to me. Or maybe it’s completely overhauled in 4.0?

**Comment by lawnjelly** [comment]
There's an extensive literature on this, see #35606, #41491, #43554 etc. Have fun.
TLDR .. we tried some fixes in core for 3.3 betas but these were reverted, as they caused problems.

**Comment by JosiahJohnson** [comment]
Good to know. Thanks for looking into it.
